pkg/controller/csi_extension_server.go

```diff
@@ -0,0 +1,128 @@
+		if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+			return nil, status.Errorf(codes.Internal, "Error creating volume group: %s", err.Error())
+		}
+	}
+	if resp.ID != "" {
+		resp, err = s.Arrays()[arr].GetClient().CreateVolumeGroupSnapshot(ctx, resp.ID, &reqParams)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error creating volume group snapshot: %s", err.Error())
+			}
+		}
+
+		volGroup, err = s.Arrays()[arr].GetClient().GetVolumeGroup(ctx, resp.ID)
+		log.Infof("%v", volGroup)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error getting volume group snapshot: %s", err.Error())
+			}
+		}
+
+		for _, v := range volGroup.Volumes {
+			var snapState bool
+			if v.State == StateReady {
+				snapState = true
+			}
+			etime, _ := time.Parse(time.RFC3339, v.CreationTimeStamp)
+			int64Time := etime.Unix()
+			snapsList = append(snapsList, &vgsext.Snapshot{
+				Name:          v.Name,
+				SnapId:        v.ID + "/" + arr + "/" + strings.Split(request.SourceVolumeIDs[0], "/")[2],
+				ReadyToUse:    snapState,
+				CapacityBytes: v.Size,
+				SourceId:      v.ProtectionData.SourceID + "/" + arr + "/" + strings.Split(request.SourceVolumeIDs[0], "/")[2],
+				CreationTime:  int64Time,
+			})
+		}
+	}
+
+	etime, _ := time.Parse(time.RFC3339, volGroup.CreationTimeStamp)
+	int64Time := etime.Unix()
+
+	return &vgsext.CreateVolumeGroupSnapshotResponse{
+		SnapshotGroupID: volGroup.ID,
+		Snapshots:       snapsList,
+		CreationTime:    int64Time,
+	}, nil
+}
+
+//validate if request has VGS name, and VGS name length < 27 chars
+func validateCreateVGSreq(request *vgsext.CreateVolumeGroupSnapshotRequest) error {
+	if request.Name == "" {
+		err := status.Error(codes.InvalidArgument, "CreateVolumeGroupSnapshotRequest needs Name to be set")
+		log.Errorf("Error from validateCreateVGSreq: %v ", err)
+		return err
+	}
+
+	//name must be less than 28 chars, because we name snapshots with -<index>, and index can at most be 3 chars
+	if len(request.Name) > 27 {
+		err := status.Errorf(codes.InvalidArgument, "Requested name %s longer than 27 character max", request.Name)
+		log.Errorf("Error from validateCreateVGSreq: %v ", err)
+		return err
+	}
+
+	if len(request.SourceVolumeIDs) == 0 {
+		err := status.Errorf(codes.InvalidArgument, "Source volumes are not present")
+		log.Errorf("Error from validateCreateVGSreq: %v ", err)
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) ParseVolumeHandle(ctx context.Context, request *vgsext.VolumeHandleRequest) (*vgsext.VolumeHandleResponse, error) {
+	if request.VolumeHandle == "" {
+		return nil, errors.New("volume handle should not be empty")
+	}
+	volHandle := strings.Split(request.VolumeHandle, "/")
+	return &vgsext.VolumeHandleResponse{
```

Comment:
check for number of elements returned by Split() before accessing the elements.

Reply:
This is removed now